proxy/http2/Http2ClientSession.cc

```diff
@@ -711,3 +848,266 @@ Http2ClientSession::add_url_to_pushed_table(const char *url, int url_len)
+  }
+}
+
+bool
+Http2ClientSession::is_state_open() const
+{
+  return session_state == &Http2ClientSession::state_open;
+}
+
+bool
+Http2ClientSession::is_state_closed() const
+{
+  return session_state == &Http2ClientSession::state_closed;
+}
+
+bool
+Http2ClientSession::is_state_readable() const
+{
+  return session_state == &Http2ClientSession::state_open || session_state == &Http2ClientSession::state_graceful_shutdown;
+}
+
+bool
+Http2ClientSession::is_state_writable() const
+{
+  return session_state == &Http2ClientSession::state_open || session_state == &Http2ClientSession::state_graceful_shutdown ||
+         session_state == &Http2ClientSession::state_goaway;
+}
+
+/**
+   HTTP2_SESSION_EVENT_REENABLE handler for state_open & state_graceful_shutdown
+ */
+int
+Http2ClientSession::_handle_ssn_reenable_event(int event, void *edata)
+{
+  ink_assert(session_state == &Http2ClientSession::state_open || session_state == &Http2ClientSession::state_graceful_shutdown);
+  ink_assert(event == HTTP2_SESSION_EVENT_REENABLE);
+
+  if (edata == nullptr) {
+    return EVENT_DONE;
+  }
+
+  Event *e = static_cast<Event *>(edata);
+
+  // VIO will be reenableed in this handler
+  int res = (this->*session_handler)(VC_EVENT_READ_READY, static_cast<VIO *>(e->cookie));
+
+  // Clear the event after calling session_handler to not reschedule REENABLE in it
+  this->_reenable_event = nullptr;
+
+  return res;
+}
+
+/**
+   VC event handler for state_open & state_graceful_shutdown
+ */
+int
+Http2ClientSession::_handle_vc_event(int event, void *edata)
+{
+  ink_assert(session_state == &Http2ClientSession::state_open || session_state == &Http2ClientSession::state_graceful_shutdown);
+
+  int res = EVENT_DONE;
+
+  switch (event) {
+  case VC_EVENT_READ_READY:
+  case VC_EVENT_READ_COMPLETE: {
+    ink_assert(edata != nullptr);
+    ink_assert(edata == read_vio);
+
+    res = (this->*session_handler)(event, edata);
+
+    break;
+  }
+  case VC_EVENT_WRITE_READY:
+  case VC_EVENT_WRITE_COMPLETE: {
+    ink_assert(edata != nullptr);
+    ink_assert(edata == write_vio);
+
+    this->connection_state.restart_streams();
+    if ((Thread::get_hrtime() >= this->_write_buffer_last_flush + HRTIME_MSECONDS(this->_write_time_threshold))) {
+      this->flush();
+    }
+
+    break;
+  }
+  case VC_EVENT_ACTIVE_TIMEOUT:
+  case VC_EVENT_INACTIVITY_TIMEOUT: {
+    dying_event = event;
+
+    REMEMBER(event, recursion);
+    Http2SsnDebug("state_goaway by %s (%d)", get_vc_event_name(event), event);
+    _set_state_goaway(Http2ErrorCode::HTTP2_ERROR_INTERNAL_ERROR);
+
+    // signal timeout event to txns
+    connection_state.cleanup_streams(event);
+
+    break;
+  }
+  case VC_EVENT_ERROR:
+  case VC_EVENT_EOS: {
+    HTTP2_SET_SESSION_STATE(&Http2ClientSession::state_aborted);
+    Http2SsnDebug("state_aborted by %s (%d)", get_vc_event_name(event), event);
+
+    dying_event = event;
+
+    // No more read/write op
+    _vc->do_io_close();
+    _vc = nullptr;
+
```

Comment:
I presume this is the immediate cause of the state_cache_open_write crash.  In the core dying_event is set to 104.  But presumably we did not get the HttpSM to stop directly as a result of the cleanup_streams() and the cache_write_open event snuck in before we could get everything shutdown.

Reply:
It looks like this is a race of callback from AIO and closing vc. Actually, I don't understand why `HttpSM::state_cache_open_write` needs the client vc. IMO, we can write to the cache without client vc.

> In the current code the _vc->do_io_close() is not called until everything is shutdown and we are getting ready to free the Http2ClientSession object.

But this is what we have now. I'll adjust this PR to follow this assumption. However, in the long term, we might want to change the design - e.g. when a client is aborted & background-fill is working, we wouldn't want to keep resources for the aborted client.

Reply:
d53ac5c get rid of this and ignore VC events on state_aborted for just in case.